File: CodeForces/src/LeetCode.java
import java.util.*;


public class LeetCode {
    static int[][] int2d(String x){
        x = x.substring(2, x.length()-2);
        String[] d2 = x.split("\\],\\[");
        int n = d2.length, m = d2[0].split(",").length;
        int[][] res = new int[n][m];
        for (int i = 0;i < n;i++){
            String[] cur = d2[i].split(",");
            for (int j = 0;j < m;j++) res[i][j] = Integer.parseInt(cur[j]);
        }
        return res;
        
    }
    static int[] int1d(String x){
        x = x.substring(1, x.length()-1);
        String[] tp = x.split(",");
        int n = tp.length;
        int[] res = new int[n];
        for (int i = 0;i < n;i++){
            res[i] = Integer.parseInt(tp[i]);
        }
        return res;
    }
    public static void main(String[] args) {
        String[] x = {"[4,7,15,8,3,5]", "[[2,1],[3,2]]", "[[18,19],[3,12],[17,19],[2,13],[7,10]]"};
        System.out.println(new Solution().findValidSplit(int1d(x[0]))
        );
    }
}

class Solution {
    long gcd(long a, long b){
        return a==0?b:gcd(b%a, a);
    }
    public int findValidSplit(int[] nums) {
        int[] cnt = new int[(int)1e6+10];
        int n = nums.length;
        for (int i = 0;i < n;i++){
            int t = nums[i];
            for (int j = 2;j*j <= t;j++){
                if (t % j == 0){
                    while (t % j == 0){
                        cnt[j]++;
                        t /= j;
                    }
                }
            }cnt[t]++;
        }
        int a = 0, b = 0;
        int[] has = new int[(int)1e6+10];
        for (int i = 0;i < n-1;i++){
            int t = nums[i];
            for (int j = 2;j*j <= t;j++){
                if (t % j == 0){
                    if (has[j]==0){
                        has[j] = 1;
                        a++;
                    }
                    while (t % j == 0){
                        if (--cnt[j] == 0) b++;
                        t /= j;
                    }
                }
            }if (a == b) return i;
        }
        return -1;
    }
}

File: PY/LeetCode.py
from math import gcd
from typing import *
from collections import defaultdict

def qpow(a, b, c):
    res = 1
    while b:
        if (b&1)==1: res = (res*a)%c
        a = (a*a)%c
        b >>= 1
    return res

class Solution:
    def findValidSplit(self, nums: List[int]) -> int:
        M = int(1e9+7)
        s, p = 1, 1
        for i in nums: s = (s*i)%M
        print(s)
        for i in range(len(nums)-1):
            p = (p*nums[i])%M
            s = s*qpow(nums[i], M-2, M)
            print(str(p)+" "+str(s))
            if gcd(p, s) == 1: return i
        return -1

s = Solution()
print(s.findValidSplit([4,7,8,15,3,5]))
    